shadowsocks/tcp.go

```diff
@@ -71,39 +54,49 @@ func debugTCP(cipherID, template string, val interface{}) {
 	}
 }
 
-func findAccessKey(clientConn onet.DuplexConn, cipherList CipherList) (string, onet.DuplexConn, []byte, error) {
-	// This must have enough space to hold the salt + 2 bytes chunk length + AEAD tag (Overhead) for any cipher
-	firstBytes := make([]byte, 0, 32+2+16)
-	// Constant of zeroes to use as the start chunk count. This must be as big as the max NonceSize() across all ciphers.
-	zeroCountBuf := [12]byte{} // MaxCountSize
+func findAccessKey(clientConn onet.DuplexConn, cipherList CipherList, m metrics.ShadowsocksMetrics) (string, onet.DuplexConn, []byte, error) {
+	clientIP := remoteIP(clientConn)
+	ciphers := cipherList.SnapshotForClientIP(clientIP)
+	firstBytes := make([]byte, tcpHeader)
+	if n, err := io.ReadFull(clientConn, firstBytes); err != nil {
+		return "", clientConn, nil, fmt.Errorf("Reading header failed after %d bytes: %v", n, err)
+	}
+
+	findStartTime := time.Now()
+	entry := findEntry(firstBytes, ciphers)
+	timeToCipher := time.Now().Sub(findStartTime)
+	if m != nil {
+		m.AddTCPCipherSearch(timeToCipher, entry != nil)
+	}
+	if entry == nil {
+		return "", clientConn, nil, fmt.Errorf("Could not find valid TCP cipher")
+	}
+
+	// Move the active cipher to the front, so that the search is quicker next time.
+	cipherList.MarkUsedByClientIP(entry, clientIP)
+	id, cipher := entry.Value.(*CipherEntry).ID, entry.Value.(*CipherEntry).Cipher
+	ssr := NewShadowsocksReader(io.MultiReader(bytes.NewReader(firstBytes), clientConn), cipher)
+	ssw := NewShadowsocksWriter(clientConn, cipher)
+	salt := firstBytes[:cipher.SaltSize()]
+	return id, onet.WrapConn(clientConn, ssr, ssw).(onet.DuplexConn), salt, nil
+}
+
+// Implements a trial decryption search.
```

Comment:
Please keep the comment that this assumes that all ciphers are AEAD.

Reply:
Done